makeMockable incompatible with DefaultSignatures

It is common to use the [DefaultSignatures][] extension to provide default implementations of class methods.  When using a class as an abstraction layer over I/O, default implementations can be provided for [MonadIO][] instances.  This reduces boilerplate when the class has more than one instance (using `MonadIO`), *and* it allows the allows the implementation to be written alongside the method signatures, making it easier to maintain.  The example in the documentation could be implemented as follows:

[DefaultSignatures]: <https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/default_signatures.html>
[MonadIO]: <https://hackage.haskell.org/package/base-4.16.0.0/docs/Control-Monad-IO-Class.html#t:MonadIO>

```haskell
class Monad m => MonadFilesystem m where
  readFile :: FilePath -> m String

  default readFile :: MonadIO m => FilePath -> m String
  readFile = liftIO . readFile

  writeFile :: FilePath -> String -> m ()

  default writeFile :: MonadIO m => FilePath -> String -> m ()
  writeFile = (liftIO .) . writeFile
```

Unfortunately, [makeMockable][] fails with an error like the following:

[makeMockable]: <https://hackage.haskell.org/package/HMock-0.5.0.0/docs/Test-HMock-TH.html#v:makeMockable>

```
/path/to/test/Spec.hs:25:1: error:
    Cannot derive MockT because Demo.MonadFileSystem has unmockable methods.
   |
25 | makeMockable [t|MonadFileSystem|]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

This error is raised when [instExtraMembers][] is [not empty][].  It is set in [makeInstance][], where `members` is processed using [getMethod][].  Perhaps support for [DefaultSignatures][] could be added by refactoring this code so that `default` members are ignored?

[instExtraMembers]: <https://github.com/cdsmith/HMock/blob/b6813a2b405eeafc2f1cb8bbcf81763f4d876cba/src/Test/HMock/TH.hs#L124>
[not empty]: <https://github.com/cdsmith/HMock/blob/b6813a2b405eeafc2f1cb8bbcf81763f4d876cba/src/Test/HMock/TH.hs#L646>
[makeInstance]: <https://github.com/cdsmith/HMock/blob/b6813a2b405eeafc2f1cb8bbcf81763f4d876cba/src/Test/HMock/TH.hs#L170>
[getMethod]: <https://github.com/cdsmith/HMock/blob/b6813a2b405eeafc2f1cb8bbcf81763f4d876cba/src/Test/HMock/TH.hs#L199>

--- a/src/Test/HMock/TH.hs
+++ b/src/Test/HMock/TH.hs
@@ -177,7 +177,7 @@ makeInstance ::
   [Dec] ->
   Q Instance
 makeInstance options ty cx tbl ps m members = do
-  processedMembers <- mapM (getMethod ty m tbl) members
+  processedMembers <- mapM (getMethod ty m tbl) $ filter isRelevantMember members
   (extraMembers, methods) <-
     partitionEithers <$> zipWithM memberOrMethod members processedMembers
   return $
@@ -190,6 +190,10 @@ makeInstance options ty cx tbl ps m members = do
         instExtraMembers = extraMembers
       }
   where
+    isRelevantMember :: Dec -> Bool
+    isRelevantMember DefaultSigD{} = False
+    isRelevantMember _ = True
+
     memberOrMethod :: Dec -> Either [String] Method -> Q (Either Dec Method)
     memberOrMethod dec (Left warnings) = do
       when (mockVerbose options) $ mapM_ reportWarning warnings

--- a/test/Classes.hs
+++ b/test/Classes.hs
@@ -1,5 +1,6 @@
 {-# LANGUAGE CPP #-}
 {-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
 {-# LANGUAGE DeriveAnyClass #-}
 {-# LANGUAGE DerivingStrategies #-}
 {-# LANGUAGE FlexibleInstances #-}
@@ -543,6 +544,32 @@ polyResultTests = describe "MonadPolyResult" $ do
       x <- polyResult (3 :: Int)
       liftIO $ x `shouldBe` 3
 
+class MonadDefaultSignatures m where
+  methodWithDefault :: Int -> m Int
+
+  default methodWithDefault :: MonadIO m => Int -> m Int
+  methodWithDefault = liftIO . methodWithDefault
+
+instance MonadDefaultSignatures IO where
+  methodWithDefault = pure
+
+makeMockable [t|MonadDefaultSignatures|]
+
+defaultSignaturesTests :: SpecWith ()
+defaultSignaturesTests = describe "MonadDefaultSignatures" $ do
+  it "generates mock impl" $
+    example $ do
+      decs <- runMockT $ do
+        allowUnexpected $ QReifyInstances_ anything anything |-> []
+        $(onReify [|expectAny|] ''MonadDefaultSignatures)
+        runQ (makeMockable [t|MonadDefaultSignatures|])
+      evaluate (rnf decs)
+  it "is mockable" $
+    example . runMockT $ do
+      expect $ MethodWithDefault_ anything |=> \MethodWithDefault{} -> return 42
+      x <- methodWithDefault 0
+      liftIO $ x `shouldBe` 42
+
 class MonadExtraneousMembers m where
   data SomeDataType m
   favoriteNumber :: SomeDataType m -> Int
@@ -812,6 +839,7 @@ classTests = describe "makeMockable" $ do
   polyArgTests
   unshowableArgTests
   monadInArgTests
+  defaultSignaturesTests
   extraneousMembersTests
   rankNTests
   defaultReturnTests

